### PR TITLE
FE-876 - Add getGroup and removeGroup methods to core controller.

### PR DESCRIPTION
**Pull request by layton-glympse**
@j5kay PTAL
@adubovsky PTAL

### CODE REVIEW

**Comment by layton-glympse** [comment]
Build is failing ATM because of other issues in the project around eslint stuff. I'm not sure how 2.6.0 was successfully built since the items have nothing to do with this code. Are we --force building now, or is there something I'm missing?

**Comment by layton-glympse** [comment]
Example usage from, say, enroute-journey:
```
var groups = adapter.core.getGroups(); // Array of groups
adapter.core.removeGroup(groups[0].getName()); // true
```

File: app/src/adapter/PublicGroupController.js [hunk 1]
```diff
@@ -89,6 +89,23 @@ define(function(require, exports, module)
 					break;
 				}
 
+				case r.removeGroup:
```
Comment (adubovsky) [inline app/src/adapter/PublicGroupController.js]: code style:
- braces around each case
- brace on next line (after `if`, `for`)

File: app/src/adapter/PublicGroupController.js [hunk 2]
```diff
@@ -89,6 +89,23 @@ define(function(require, exports, module)
...
+					if (!groups.hasOwnProperty(args)) {
+						return false;
+					}
+					return delete groups[args];
```
Comment (adubovsky) [inline app/src/adapter/PublicGroupController.js]: We should also care about `pollGroups` array (which actually looks same as values of `groups` object.

**Comment by j5kay** [comment]
I built 2.6.0 without issue. There are some warnings, but nothing that are considered errors. What are you guys seeing?

**Comment by j5kay** [comment]
re: Code style -- @layton-glympse: before generating a pull request and/or committing updates to an existing PR, you need to review your changes via a diff tool that allows you to see the actual updates (i.e. Tower and/or github), as your current workflow doesn't seem to be catching these simple deviations.

This PR has numerous style differences (spaces vs. tabs, braces, etc.) that have been a consistent problem for projects outside of your preferred coding style. Please take extra time in the future when generating updates for others to review.

**Comment by j5kay** [comment]
As for removing an invite from the viewer, there is a `removeInvites` interface on the map (see https://github.com/Glympse/web-viewer-gmaps/blob/develop/VIEWER-API.md, about a quarter of the way down, for details). It should remove everything related to the associated InviteView object.

**Comment by layton-glympse** [comment]
@j5kay I have been reviewing on GitHub before asking for PR. I'm not sure what's up but I'll take another pass on why my editor keeps doing this.

File: app/src/adapter/PublicGroupController.js [hunk 3]
```diff
@@ -89,6 +89,45 @@ define(function(require, exports, module)
...
+					{
+						if (pollGroups[i].getName() === args)
+						{
+							pollGroups.splice(i, 1);
+							break;
+						}
+					}
+
+					controller.notify(m.GroupStatus, {
+						status: 'success',
```
Comment (adubovsky) [inline app/src/adapter/PublicGroupController.js]: `status` should be boolean as of current impl.

**Comment by j5kay** [comment]
Added some updates:
- Lots of input error checking added
- Fix RAF timer issues
- Add support for getting a single group, or an array list of groups
- toLowerCase() everything to prevent casing issues (which are irrelevant for PublicGroups)
- Fix data return value handling
- Clear polling timer if all groups are removed
- Give option to also remove invites on the map from a group that is removed (default is still to leave the invites on the map)
- Add support for sub-controllers to issue commands to other sub-controllers
- Small hack to allow for sending a command from core to map controller
- Add removeGroup option to test-server page
- Still need to add docs + test options for getGroups API call

File: app/src/adapter/PublicGroupController.js [hunk 4]
```diff
@@ -79,16 +79,87 @@ define(function(require, exports, module)
...
+								break;
+							}
+						}
+
+						invites = groups[groupName].getInvites();
+
+						// Optionally remove invites from the map as well
+						if (removeInvites)
+						{
+							setTimeout(function()
```
Comment (adubovsky) [inline app/src/adapter/PublicGroupController.js]: Why do we not use `raf` here?

File: app/src/adapter/PublicGroupController.js [hunk 5]
```diff
@@ -79,16 +79,87 @@ define(function(require, exports, module)
...
+							}
+						}
+
+						args = newArr;
+					}
+
+					for (var key in groups)
+					{
+						if (groups.hasOwnProperty(key) &&
+						   (!args || (typeof args === 'string' && args.toLowerCase() === key) || (Array.isArray(args) && args.indexOf(key) >= 0))
```
Comment (adubovsky) [inline app/src/adapter/PublicGroupController.js]: Better to store the result of `Array.isArray(args)` after the first call above and use it here. Or just use `|| newArr.indexOf(key)`

Also this method is IE9+, is it OK? )

File: app/src/adapter/PublicGroupController.js [hunk 6]
```diff
@@ -127,10 +198,23 @@ define(function(require, exports, module)
 				header = groupInfo;
 			}
 
+			if ((name === null || name === undefined) && !header)
```
Comment (adubovsky) [inline app/src/adapter/PublicGroupController.js]: Can group name be empty? Probably can use just `!name`.
Also do we need to check for header additionally here? If the header will be defined but w/o name it's also invalid case, isn't it?

Comment (j5kay) [inline app/src/adapter/PublicGroupController.js]: Reduced down. I was a little worried about other possible/no names if the header is given, but the rest of the logic depends on a non-empty string for the group name, so it should reject the input if this is the case.

File: app/src/adapter/PublicGroupController.js (hunk 5 above)
Comment (j5kay) [inline app/src/adapter/PublicGroupController.js]: Polyfilled isArray and lightened list args option check

File: app/src/adapter/PublicGroupController.js (hunk 4 above)
Comment (j5kay) [inline app/src/adapter/PublicGroupController.js]: Done!

**Comment by adubovsky** [comment]
LGTM

**Pull request merged**